examples/tutorials/mnist_pytorch/idle_gpu_watcher.py

```diff
@@ -0,0 +1,174 @@
+                sys.exit(1)
+            if self._debug:
+                print(f"DEBUG: {gpu_uuid} usage {gpu_util}% ")
+
+            self._gpu_util_samples[gpu_uuid].popleft()
+
+    def watch_process(self, args):
+        if args is None or len(args) == 0:
+            raise ValueError("args cannot be emtpy")
+
+        p = subprocess.Popen(args)
+        print("IDLE_GPU_WATHCER: "
+              f"Starting idle GPU watcher: min_threshhold={self._min_threshhold}%, "
+              f"sample_freq={self._sample_freq}, num_samples={self._num_samples}, "
+              f"delay_samples={self._delay_samples}, window_size={self._full_window}")
+
+        try:
+            while True:
+                if p.poll() is None:
+                    time.sleep(self._sample_freq)
+
+                    if self._delay_samples > 0:
+                        self._delay_samples -= 1
+                        continue
+
+                    self._check_idle()
+                    continue
+
+                break
+
+        finally:
+            if p.poll() is None:
+                print("IDLE_GPU_WATCHER: Training process still alive, killing.")
+                p.kill()  # SIGKILL because it is likely hung
+
+        return p.wait()
+
+
+class AParser(argparse.ArgumentParser):
+    def error(self, message):
+        self.print_help()
+        sys.stderr.write(f"\nerror: {message}\n")
+        sys.exit(2)
+
+
+if __name__ == "__main__":
+    """
+    This is for Determined >= 0.17.7.
+
+    To use, add the following to your experiment config:
+
+entrypoint:
+  - "python3"
+  - "-m"
+  - "idle_gpu_watcher"
+  - "5"       # Minimum GPU usage which triggers a failure.
+  - "5"       # How frequently to sample the GPU usage.
+  - "10"      # Number of samples to look at when evaluating GPU usage.
+  - "6"       # Number of samples to wait before evaluating GPU usage.
+  - "true"    # Print debug statements
+  - "python3"
+  - "-m"
+  - "determined.launch.autohorovod"
+  - "model_def:Trial"
+    """
+
+    parser = AParser(description="Launch a process and record GPU usage every sample_freq seconds. "
+                     "Exit if at least num_samples are below min_percentage after delay_samples has"
+                     " passed.")
+
+    parser.add_argument("min_percentage", type=int, default=5,
+                        help="Minimum GPU usage which triggers a failure")
+    parser.add_argument("sample_freq", type=int, default=5,
+                        help="How frequently to sample each GPU usage")
+    parser.add_argument("num_samples", type=int, default=12,
+                        help="Number of samples to look at when evaluating GPU usage")
+    parser.add_argument("delay_samples", type=int, default=12,
+                        help="Number of samples to wait before evaluating GPU usage")
+    parser.add_argument("debug", type=bool, default=True,
+                        help="Display debug information")
```

Comment:
recommend:
```python
    # make these optional flags, so the cli is more readable
    parser.add_argument("--min_percentage", type=int, default=5,
                        help="Minimum GPU usage which triggers a failure")
    parser.add_argument("--sample_freq", type=int, default=5,
                        help="How frequently to sample each GPU usage")
    parser.add_argument("--num_samples", type=int, default=12,
                        help="Number of samples to look at when evaluating GPU usage")
    parser.add_argument("--delay_samples", type=int, default=12,
                        help="Number of samples to wait before evaluating GPU usage")
    
    # use flag with action="store_true" to do what `--debug` would normally do
    parser.add_argument("--debug", action="store_true",
                        help="Display debug information")
```

Then change the config to:
```yaml
entrypoint:
  - python3
  - -m
  - idle_gpu_watcher
  - --min-percentage
  - "5"
  - --sample-freq
  - "5"
  - --sample-window
  - "10"
  - --startup-delay-samples
  - "6"
  - --debug
  - python3 
  - -m
  - determined.launch.autohorovod
  - model_def:MNistTrial
```

- I took out the unnecessary double-quotes
- I recommend using --sample-window instead of --num-samples, since "window" is a very common concept when writing filters in DSP, which is sort of what we're doing)
- I recommend using --startup-delay-samples instead of --delay-samples because it becomes totally clear what it means

Reply:
Done! Good suggestions.